record/record_test.go

```diff
@@ -452,3 +486,70 @@ func TestComplexRecover(t *testing.T) {
 		t.Fatal("Unexpected output in r2's data")
 	}
 }
+
+func TestMuliBlockCorruptionRecover(t *testing.T) {
+	records, buf, err := generateTestRecords(
+		// The first record will consume 3 entire blocks but a fraction of the 4th.
+		blockSize*3,
+		// The second record will completely fill the remainder of the 4th block.
+		3*(blockSize-headerSize)-2*blockSize-2*headerSize,
+		// Consume the entirety of the 5th block.
+		blockSize-headerSize,
+		// Consume the entirety of the 6th block.
+		blockSize-headerSize,
+		// Consume roughly half of the 7th block.
+		blockSize/2,
+	)
+	if err != nil {
+		t.Fatalf("generateTestRecords: %v", err)
+	}
+
+	// Corrupt the checksum for the portion of the first record that exists in the 4th block.
+	buf[blockSize*3+0] = 0x00
+	buf[blockSize*3+1] = 0x00
+	buf[blockSize*3+2] = 0x00
+	buf[blockSize*3+3] = 0x00
+
+	// Now corrupt the two blocks in a row that correspond to records[2] and records[3].
+	buf[blockSize*4+0] = 0x00
+	buf[blockSize*4+1] = 0x00
+	buf[blockSize*4+2] = 0x00
+	buf[blockSize*4+3] = 0x00
+	buf[blockSize*5+0] = 0x00
+	buf[blockSize*5+1] = 0x00
+	buf[blockSize*5+2] = 0x00
+	buf[blockSize*5+3] = 0x00
+
+	// The first record should fail, but only when we read deeper beyond the first block.
+	r := NewReader(bytes.NewReader(buf))
+	r0, err := r.Next()
+	if err != nil {
+		t.Fatalf("Next: %v", err)
+	}
+
+	// Reading deeper should yield a checksum mismatch.
+	_, err = ioutil.ReadAll(r0)
+	if err == nil {
+		t.Fatal("Exptected a checksum mismatch error, got nil")
+	}
+	if !strings.Contains(err.Error(), "checksum mismatch") {
+		t.Fatalf("Unexpected error returned: %v", err)
+	}
+
+	// Recover from that checksum mismatch.
+	r.Recover()
+
+	// All of the data in the second record is lost because the first
+	// record shared a partial block with it. The following two records
+	// have corrupted checksums as well, so the call above to r.Recover
+	// should result in r.Next() being a reader to the 5th record.
+	r4, err := r.Next()
+	if err != nil {
+		t.Fatalf("Next: %v", err)
+	}
+
+	r4Data, _ := ioutil.ReadAll(r4)
+	if !bytes.Equal(r4Data, records[4]) {
+		t.Fatal("Unexpected output in r4's data")
+	}
+}
```

Comment:
Can I get a test where the last block is corrupted, with and without it being a full blockSize-sized block?